File: Code/Websites/DanpheEMR/wwwroot/DanpheApp/src/app/accounting/transactions/shared/transaction.model.ts
import {
    FormBuilder,
    FormControl,
    FormGroup,
    Validators
} from '@angular/forms';
import * as moment from 'moment/moment';
import { ENUM_DateTimeFormat } from '../../../shared/shared-enums';
import { BillingAccountingSyncModel } from './billing-accounting-sync.model';
import { TransactionItem } from './transaction-item.model';

export class TransactionModel {
    public TransactionId: number = 0;
    public VoucherId: number = 0;
    public FiscalYearId: number = 0;
    public TransactionDate: string = null;
    public CreatedBy: number = 0;
    public CreatedOn: string = null;
    public IsActive: boolean = true;
    public IsBackDateEntry: boolean = false;
    public Remarks: string = "";
    public PayeeName: string = "";
    public ChequeNumber: number = 0;
    //public ReferenceTransactionId: number = 0;
    public SectionId: number = 0;
    public VoucherNumber: string = "";
    public TotalAmount: number = 0;
    public TransactionItems: Array<TransactionItem> = new Array<TransactionItem>();
    //public TransactionLinks: Array<TransactionLink> = new Array<TransactionLink>();
    public BillSyncs: Array<BillingAccountingSyncModel> = new Array<BillingAccountingSyncModel>();
    public TransactionValidator: FormGroup = null;
    public BillingAccountingSyncIds: string = "";
    //only used in client side
    public RefTxnVoucherNumber: number = 0;
    // public CostCenterId: number = 0; //Dev : 26 Jan, 2023 Replaced VoucherHeadId to CostCenterId : DevN : 13th April 23 Costcenter is moved into item level...
    public TransactionType: string = "";
    public TUId: number = 0;
    public DayVoucherNumber: number = 0;

    public ModifiedBy: number = 0;
    public ModifiedOn: string = null;
    public Reason: string = null;
    public ChequeDate: string = null;
    public VoucherSerialNo: number = 0;
    public IsVoucherReversed: boolean = false;
    constructor() {
        this.TransactionDate = moment().format(ENUM_DateTimeFormat.Year_Month_Day_Hour_Minute);
        this.CreatedOn = moment().format(ENUM_DateTimeFormat.Year_Month_Day_Hour_Minute);
        var _formBuilder = new FormBuilder();
        this.TransactionValidator = _formBuilder.group({
            'VoucherId': ['', Validators.compose([])],
            'PayeeName': ['', Validators.required],
            'ChequeNumber': ['', Validators.compose([Validators.required, Validators.pattern('^[0-9]+$')])],
            'Remarks': ['', Validators.compose([Validators.maxLength(500)])],
            'RefTxnVoucherNumber': ['', Validators.compose([])],
            //'CostCenter': ['', Validators.required],
            'TransactionDate': ['', Validators.compose([Validators.required, this.dateValidators])]
        });
    }

    public IsDirty(fieldName): boolean {
        if (fieldName == undefined)
            return this.TransactionValidator.dirty;
        else
            return this.TransactionValidator.controls[fieldName].dirty;
    }

    public IsValid(): boolean { if (this.TransactionValidator.valid) { return true; } else { return false; } }
    public IsValidCheck(fieldName, validator): boolean {
        if (fieldName == undefined) {
            return this.TransactionValidator.valid;

        }
        else
            return !(this.TransactionValidator.hasError(validator, fieldName));
    }

    dateValidators(control: FormControl): { [key: string]: boolean } {

        //get current date, month and time
        var currDate = moment().format('YYYY-MM-DD');

        //if positive then selected date is of future else it of the past
        if ((moment(control.value).diff(currDate) > 0) ||
            (moment(control.value).diff(currDate, 'years') < -200)) // this will not allow the age diff more than 200 is past
            return { 'wrongDate': true };
    }
    public UpdateValidator(onOff: string, formControlName: string, validatorType: string) {
        let validator = null;
        if (validatorType == 'required' && onOff == "on") {
            validator = Validators.compose([Validators.required]);
        }
        else {
            validator = Validators.compose([]);
        }
        this.TransactionValidator.controls[formControlName].validator = validator;
        this.TransactionValidator.controls[formControlName].updateValueAndValidity();
    }

}
export class TransactionViewModel {
    public TransactionId: number = 0;
    public VoucherName: string = null;
    public FiscalYearName: string = null;
    public TransactionDate: string = null;
    public Remarks: string = null;
    public IsBackDateEntry: boolean = false;
    public VoucherType: string = null;
    public IsGroupTxn: boolean = true;
    public IsEditable: boolean = null;
    public VoucherNumber: number = 0;
    public IsVerified: boolean = false;
    public IsVoucherReversed: boolean = false;
    public VerifiedBy: string = "";
    public Preparedby: string = "";
    public IsAllowReverseVoucher: boolean = false;
    public ChequeNumber: string = "";
    public PayeeName: string = "";
    public FiscalYear: string = "";
    public TransactionItems: Array<{
        LedgerGroupName: string,
        LedgerName: string,
        Name: string,
        DrCr: boolean,
        Amount: number,
        Code: string,
        Details: any[],
        SupplierDetails: any[],
        TransactionType: any[],
        Remarks: string,
        VoucherNumber: any[],
        SubLedgers: any[]
        CostCenterId: number,
        CostCenterName: string,
        Description: string
    }> = []
    public HospitalId: number = 0;
}
//export class BillingItemsUpdateModel {
//    public BillingTransactionItemId: number = 0;
//    public IsTransferredToACC: number = 0;
//}

File: Code/Components/DanpheEMR.ServerModel/AccountingModels/Transactions/TransactionModel.cs

using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DanpheEMR.ServerModel
{
    public class TransactionModel
    {
        [Key]
        public int TransactionId { get; set; }
        public int VoucherId { get; set; }
        public int FiscalyearId { get; set; }
        public string Remarks { get; set; }
        public DateTime TransactionDate { get; set; }
        public DateTime CreatedOn { get; set; }
        public int CreatedBy { get; set; }
        public bool IsActive { get; set; }
        //public int CostCenterId { get; set; } // Dev : 26 Jan 2023 Replaced VoucherHeadId into CostCenterId : Devn 13th April 23 Moved CostCenter in item level..
        //public int? ReferenceTransactionId { get; set; }
        public string VoucherNumber { get; set; }
        public string PayeeName { get; set; }
        public string ChequeNumber { get; set; }
        public int SectionId { get; set; }
        public virtual List<TransactionItemModel> TransactionItems { get; set; }
        //public virtual List<TransactionLinkModel> TransactionLinks { get; set; }
        [NotMapped]
        public string BillingAccountingSyncIds { get; set; }
        //[NotMapped]
        //public string? ReferenceIds { get; set; }
        [NotMapped]
        public List<SyncBillingAccountingModel> BillSyncs { get; set; }
        public bool IsBackDateEntry { get; set; }
        public string TransactionType { get; set; }
        public int TUId { get; set; }
        public int DayVoucherNumber { get; set; }
        public bool IsCustomVoucher { get; set; }
        public bool IsReverseTxnAllow { get; set; }
        public bool IsEditable { get; set; }
        public DateTime? ModifiedOn { get; set; }
        public int? ModifiedBy { get; set; }
        public bool IsGroupTxn { get; set; }

        //sud-nagesh: 20June'20-- for Acc-Hospital Separation.
        public int HospitalId { get; set; }
        [NotMapped]
        public string Reason { get; set; }
        public int VoucherSerialNo { get; set; }
        public bool IsAllowReverseVoucher { get; set; }
        [NotMapped]
        public bool? IsReverseVoucher { get; set; }
        [NotMapped]
        public int? PrevTransactionId { get; set; }
        public DateTime? ChequeDate { get; set; }
        public string Status { get; set; }
        public bool IsVerified { get; set; }
        public int? VerifiedBy { get; set; }
        public DateTime? VerifiedOn { get; set; }
        public string VerificationRemarks { get; set; }
        public int? CancelledBy { get; set; }
        public DateTime? CancelledOn { get; set; }
        public string CancelledRemarks { get; set; }
        public bool IsVoucherReversed { get; set; }
    }

}
